site/_posts/2020-11-15-how-to-securely-implement-oauth-rails.md

```diff
@@ -0,0 +1,382 @@
+  # Invalidate session with FusionAuth
+  @oauth_client.request(:get, 'oauth2/logout')
+
+  # Reset Rails session
+  reset_session
+
+  redirect_to root_path
+end
+```
+
+### Polishing up our Rails session
+Now that we have the functionality to authenticate users as well as log them out of FusionAuth, we will want to complete the same cycle on our Rails app.
+
+There are numerous ways to handle user sessions in Rails, but for this example, I decided to create helper methods accessible to all sub-classes of ApplicationController.
+
+`current_user` looks for a user JWT on the session. If it exists and the email has been verified, it retrieves the user's email address from the JWT.
+If it does not exist, this means the user has logged out or their session has expired.
+
+```ruby
+# app/controllers/application_controller.rb
+
+class ApplicationController < ActionController::Base
+  helper_method :current_user
+  helper_method :logged_in?
+
+  def current_user
+    if session[:user_jwt]
+      token = session[:user_jwt]["value"].first
+
+      if token && token["email_verified"]
+        @email = token["email"]
+      else
+        head :forbidden
+        return
+      end
+    end
+  end
+
+  def logged_in?
+    current_user.present?
+  end
+
+end
+```
+
+## Putting it all together
+
+It's time to make our Ruby on Rails OAuth flow a reality by walking through an example login. 
+Before starting up the Rails server, be sure to set `HMAC_SECRET` with the "default signing key" value in your environment making it accessible to our development configuration. 
+
+Kick the tires and light the fires!
+```
+rails s
+```
+
+Navigate to `http://localhost:3000`. Substitute the corresponding port if you are using something other than 3000.
+
+{% include _image.liquid src="/assets/img/blogs/fusionauth-example-rails/home.png" alt="The example Home screen for the Rails application." class="img-fluid" figure=false %}
+
+Log-in to the application with the user we created and registered in FusionAuth.
+
+{% include _image.liquid src="/assets/img/blogs/fusionauth-example-rails/fa-login.png" alt="The FusionAuth login page." class="img-fluid" figure=false %}
+
+We made it! We are logged in and our application knows who we are.
+
+{% include _image.liquid src="/assets/img/blogs/fusionauth-example-rails/logged-in.png" alt="The example Home screen when logged-in to the Rails application." class="img-fluid" figure=false %}
+
+Let us now assume that we are now done interacting with the application and are ready to log out by navigating to the corresponding URL.
+
+{% include _image.liquid src="/assets/img/blogs/fusionauth-example-rails/logging-out.png" alt="The FusionAuth logging out page." class="img-fluid" figure=false %}
+
+## Next steps
+Now that we have the OAuth flow working, our foundation is set to expand as needed. Here are a few ideas to tackle next:
+* Add and assign user roles such that functionality in our Rails application can be shown or hidden accordingly.
```

Comment:
Could add links to the various docs about these additional pieces of work if you want to.